File: React App/src/Task-4.js
// src/Task-4.js
import React, { useState, useCallback, useEffect } from 'react';
import Plot from 'react-plotly.js';

const g = 9.81; // Default gravity

const optimalAngle = (u, h, g) => {
    return Math.asin(1 / Math.sqrt(2 + 2 * g * h / (u ** 2))) * (180 / Math.PI); // Convert to degrees
};

const trajectory = (u, theta, h, g, numPoints = 500) => {
    const thetaRad = theta * (Math.PI / 180); // Convert to radians
    const tFlight = (u * Math.sin(thetaRad) + Math.sqrt((u * Math.sin(thetaRad)) ** 2 + 2 * g * h)) / g;
    const t = Array.from({ length: numPoints }, (_, i) => (i / (numPoints - 1)) * tFlight);
    const x = t.map(ti => u * Math.cos(thetaRad) * ti);
    const y = t.map(ti => h + u * Math.sin(thetaRad) * ti - 0.5 * g * ti ** 2);
    return { x, y };
};

const plotTrajectories = (u, h, g, givenTheta) => {
    const thetaOptimal = optimalAngle(u, h, g);
    const { x: xGiven, y: yGiven } = trajectory(u, givenTheta, h, g);
    const { x: xOptimal, y: yOptimal } = trajectory(u, thetaOptimal, h, g);
    return [
        {
            x: xGiven,
            y: yGiven,
            type: 'scatter',
            mode: 'lines',
            line: { color: 'blue' },
            name: `Given Angle: ${givenTheta.toFixed(2)} degrees`,
        },
        {
            x: xOptimal,
            y: yOptimal,
            type: 'scatter',
            mode: 'lines',
            line: { color: 'red', dash: 'dash' },
            name: `Optimal Angle: ${thetaOptimal.toFixed(2)} degrees`,
        }
    ];
};

const Task4 = () => {
    const [angle, setAngle] = useState(60);
    const [speed, setSpeed] = useState(10);
    const [height, setHeight] = useState(2);
    const [gravity, setGravity] = useState(9.81);

    const calculateData = useCallback(() => {
        return plotTrajectories(speed, height, gravity, angle);
    }, [angle, speed, height, gravity]);

    const data = calculateData();

    return (
        <div>
            <h1>Projectile Trajectories</h1>
            <Plot
                data={data}
                layout={{
                    title: 'Projectile Trajectories',
                    xaxis: { title: 'Horizontal Distance (m)' },
                    yaxis: { title: 'Vertical Distance (m)' },
                }}
                style={{ width: '100%', height: '500px' }}
            />
            <div>
                <label>
                    Angle (degrees):
                    <input
                        type="range"
                        min="-90"
                        max="90"
                        step="0.01"
                        value={angle}
                        onChange={(e) => setAngle(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="-90"
                        max="90"
                        step="0.01"
                        value={angle.toFixed(2)}
                        onChange={(e) => setAngle(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Initial Speed (m/s):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={speed}
                        onChange={(e) => setSpeed(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={speed.toFixed(2)}
                        onChange={(e) => setSpeed(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Initial Height (m):
                    <input
                        type="range"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={height}
                        onChange={(e) => setHeight(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="1000"
                        step="0.01"
                        value={height.toFixed(2)}
                        onChange={(e) => setHeight(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
            <div>
                <label>
                    Gravity (m/s²):
                    <input
                        type="range"
                        min="0"
                        max="30"
                        step="0.01"
                        value={gravity}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                    />
                    <input
                        type="number"
                        min="0"
                        max="30"
                        step="0.01"
                        value={gravity.toFixed(2)}
                        onChange={(e) => setGravity(parseFloat(e.target.value))}
                        style={{ width: '100px', marginLeft: '10px' }}
                    />
                </label>
            </div>
        </div>
    );
};

export default Task4;

File: Tasks in Python/Code/Task_4.py
# Required libraries
import numpy as np
import matplotlib.pyplot as plt

# Function to calculate the optimal angle for maximum range
def optimal_angle(u, h, g):
    return np.arcsin(1 / np.sqrt(2 + 2 * g * h / u ** 2))

# Function to calculate the range of the trajectory for a given angle and initial height
def range_of_trajectory(u, theta, h, g):
    theta_rad = np.radians(theta)
    return (u ** 2 / g) * (
        np.sin(2 * theta_rad) / (np.cos(theta_rad) + np.sqrt(np.sin(theta_rad) ** 2 + (2 * g * h) / (u ** 2)))
    )

# Function to calculate the maximum possible range
def range_maximum(u, h, g):
    theta_max = optimal_angle(u, h, g)
    return (u ** 2 / g) * np.sqrt(1 + (2 * g * h) / (u ** 2))

# Function to generate the trajectory data points for a given angle
def trajectory(u, theta, h, g):
    theta_rad = np.radians(theta)
    t_flight = (u * np.sin(theta_rad) + np.sqrt((u * np.sin(theta_rad)) ** 2 + 2 * g * h)) / g  # Total flight time
    t = np.linspace(0, t_flight, num=500)  # Time array
    x = u * np.cos(theta_rad) * t  # Horizontal distance as a function of time
    y = h + u * np.sin(theta_rad) * t - 0.5 * g * t ** 2  # Vertical distance as a function of time
    return x, y

# Function to plot the trajectories for both the given angle and the optimal angle
def plot_trajectories(u, h, g, given_theta):
    theta_optimal = np.degrees(optimal_angle(u, h, g))  # Calculate the optimal angle in degrees
    x_given, y_given = trajectory(u, given_theta, h, g)  # Trajectory for the given angle
    x_optimal, y_optimal = trajectory(u, theta_optimal, h, g)  # Trajectory for the optimal angle

    plt.figure(figsize=(12, 6))
    plt.plot(x_given, y_given, label=f'Given Angle: {given_theta} degrees', color='blue')  # Plot given angle trajectory
    plt.plot(x_optimal, y_optimal, label=f'Optimal Angle: {theta_optimal:.2f} degrees', color='red', linestyle='--')  # Plot optimal angle trajectory
    plt.title('Projectile Trajectories')  # Title of the plot
    plt.xlabel('Horizontal Distance (m)')  # Label for the x-axis
    plt.ylabel('Vertical Distance (m)')  # Label for the y-axis
    plt.legend()  # Display the legend
    plt.grid(True)  # Display a grid
    plt.show()  # Show the plot

# Parameters
u = 10             # Initial speed (m/s)
h = 2              # Initial height (m)
g = 9.81           # Acceleration due to gravity (m/s^2)
given_theta = 60   # Given launch angle (degrees)

# Plot the trajectories for the given and optimal angles
plot_trajectories(u, h, g, given_theta)
